File: backend/src/policy/policyGuard.js
/**
 * policyGuard.js — Deterministic policy enforcement.
 *
 * CRITICAL: This file contains ZERO LLM calls. Pure deterministic logic.
 * This is the most important design decision in the entire project:
 * an LLM should never be the thing that directly decides "yes, spend the money."
 *
 * Rules are applied in a specific order (order matters — first match wins).
 */

// ─── Named policy constants (easy to find and adjust) ─────────────
const MAX_ATTEMPTS = 3;
const MIN_RECOVERABLE_AMOUNT = 100; // rupees
const AUTO_APPROVE_DISCOUNT_MAX_PCT = 0; // 0% = no discount can auto-approve
const APPROVAL_DISCOUNT_MAX_PCT = 5;     // 1-5% needs merchant approval
const MIN_CONFIDENCE_AUTO_APPROVE = 0.6;

/**
 * Check whether a recovery plan passes the policy rules.
 *
 * Discount tiers:
 *   0%  → can auto-approve (if confidence is high enough)
 *   1-5% → needs merchant approval
 *   >5% → rejected outright, automatic no-discount fallback
 *
 * @param {object} caseData - The case document from MongoDB
 * @param {object} plan - The recovery plan from the Planner Agent
 * @returns {object} { decision, reason, fallback_action? }
 */
function checkPolicy(caseData, plan) {
  // Rule 1: Consent check (safety-first)
  if (caseData.has_recovery_consent === false) {
    return {
      decision: 'BLOCKED_NO_CONSENT',
      reason: 'Customer has not consented to recovery contact'
    };
  }

  // Rule 2: Retry/contact cap check (protects customer from repeated contact)
  if (caseData.attempt_number >= MAX_ATTEMPTS || caseData.contact_count >= caseData.max_contact_count) {
    return {
      decision: 'BLOCKED_STOP_RULE',
      reason: `Retry/contact cap reached (attempt ${caseData.attempt_number}/${MAX_ATTEMPTS}, contacts ${caseData.contact_count}/${caseData.max_contact_count})`
    };
  }

  // Rule 3: Minimum amount threshold (not worth the cost)
  if (caseData.amount < MIN_RECOVERABLE_AMOUNT) {
    return {
      decision: 'BLOCKED_STOP_RULE',
      reason: `Amount ₹${caseData.amount} below minimum recovery threshold (₹${MIN_RECOVERABLE_AMOUNT})`
    };
  }

  // Rule 4: Discount exceeds hard ceiling → REJECTED with automatic fallback
  if (plan.discount_requested_pct > APPROVAL_DISCOUNT_MAX_PCT) {
    return {
      decision: 'REJECTED_FALLBACK',
      reason: `Discount ${plan.discount_requested_pct}% exceeds maximum allowable ceiling (${APPROVAL_DISCOUNT_MAX_PCT}%)`,
      fallback_action: 'CREATE_PAYMENT_LINK'
    };
  }

  // Rule 5: Any discount (1-5%) requires merchant approval
  if (plan.discount_requested_pct > AUTO_APPROVE_DISCOUNT_MAX_PCT &&
      plan.discount_requested_pct <= APPROVAL_DISCOUNT_MAX_PCT) {
    return {
      decision: 'NEEDS_MERCHANT_APPROVAL',
      reason: `Any discount (${plan.discount_requested_pct}%, range 1-${APPROVAL_DISCOUNT_MAX_PCT}%) requires merchant approval`
    };
  }

  // Rule 6: Low confidence requires merchant review
  if (plan.confidence < MIN_CONFIDENCE_AUTO_APPROVE) {
    return {
      decision: 'NEEDS_MERCHANT_APPROVAL',
      reason: `Planner confidence ${plan.confidence.toFixed(2)} below auto-approve threshold (${MIN_CONFIDENCE_AUTO_APPROVE})`
    };
  }

  // Rule 7: All checks passed
  return {
    decision: 'APPROVED',
    reason: 'Within policy limits'
  };
}

module.exports = {
  checkPolicy,
  // Export constants for testing and reference
  POLICY_CONSTANTS: {
    MAX_ATTEMPTS,
    MIN_RECOVERABLE_AMOUNT,
    AUTO_APPROVE_DISCOUNT_MAX_PCT,
    APPROVAL_DISCOUNT_MAX_PCT,
    MIN_CONFIDENCE_AUTO_APPROVE
  }
};

File: backend/app/policy/policy_guard.py
from app.models.case import Case, Plan, PolicyCheck, PolicyDecision

# ─── Runtime-Mutable Configuration ──────────────────────────────────────────
# These values can be hot-swapped at runtime via apply_thresholds() without
# restarting the server. The factory defaults are preserved for revert.

_FACTORY_DEFAULTS = {
    "max_attempts": 3,
    "min_recoverable_amount": 100,       # rupees
    "auto_approve_discount_max_pct": 0,
    "approval_discount_max_pct": 5,
    "min_confidence_auto_approve": 0.6,
}

_active_config = {**_FACTORY_DEFAULTS}
_previous_config = None  # Stored on apply for one-step revert
_applied_at = None       # ISO timestamp of last apply, None = factory defaults


def get_active_thresholds() -> dict:
    """Return the currently active threshold configuration."""
    return {
        **_active_config,
        "is_factory_defaults": _active_config == _FACTORY_DEFAULTS,
        "applied_at": _applied_at,
    }


def get_factory_defaults() -> dict:
    """Return the immutable factory default thresholds."""
    return {**_FACTORY_DEFAULTS}


def apply_thresholds(thresholds: dict) -> dict:
    """
    Overwrite the active policy config with new thresholds.
    Stores the previous config for one-step revert.
    Only updates keys that exist in the config; ignores unknown keys.
    """
    global _active_config, _previous_config, _applied_at
    from datetime import datetime, timezone

    _previous_config = {**_active_config}
    for key in _FACTORY_DEFAULTS:
        if key in thresholds:
            _active_config[key] = thresholds[key]

    _applied_at = datetime.now(timezone.utc).isoformat()
    return get_active_thresholds()


def revert_thresholds() -> dict:
    """
    Revert to the previous config (if one exists), otherwise to factory defaults.
    """
    global _active_config, _previous_config, _applied_at

    if _previous_config is not None:
        _active_config = {**_previous_config}
        _previous_config = None
    else:
        _active_config = {**_FACTORY_DEFAULTS}

    _applied_at = None
    return get_active_thresholds()


# ─── Policy Check ────────────────────────────────────────────────────────────

def check_policy(case_data: Case, plan: Plan) -> PolicyCheck:
    """
    Evaluate a case + plan against the currently active policy thresholds.
    """
    cfg = _active_config

    MAX_ATTEMPTS = cfg["max_attempts"]
    MIN_RECOVERABLE_AMOUNT = cfg["min_recoverable_amount"]
    AUTO_APPROVE_DISCOUNT_MAX_PCT = cfg["auto_approve_discount_max_pct"]
    APPROVAL_DISCOUNT_MAX_PCT = cfg["approval_discount_max_pct"]
    MIN_CONFIDENCE_AUTO_APPROVE = cfg["min_confidence_auto_approve"]

    # Rule 1: Consent check
    if case_data.has_recovery_consent is False:
        return PolicyCheck(
            decision=PolicyDecision.BLOCKED_NO_CONSENT,
            reason='Customer has not consented to recovery contact'
        )
    
    # Rule 2: Retry/contact cap check
    if case_data.attempt_number >= MAX_ATTEMPTS or case_data.contact_count >= case_data.max_contact_count:
        return PolicyCheck(
            decision=PolicyDecision.BLOCKED_STOP_RULE,
            reason=f'Retry/contact cap reached (attempt {case_data.attempt_number}/{MAX_ATTEMPTS}, contacts {case_data.contact_count}/{case_data.max_contact_count})'
        )

    # Rule 3: Minimum amount threshold
    if case_data.amount < MIN_RECOVERABLE_AMOUNT:
        return PolicyCheck(
            decision=PolicyDecision.BLOCKED_STOP_RULE,
            reason=f'Amount ₹{case_data.amount} below minimum recovery threshold (₹{MIN_RECOVERABLE_AMOUNT})'
        )

    # Rule 4: Discount exceeds hard ceiling -> REJECTED with fallback
    discount = plan.discount_requested_pct or 0.0
    if discount > APPROVAL_DISCOUNT_MAX_PCT:
        return PolicyCheck(
            decision=PolicyDecision.REJECTED_FALLBACK,
            reason=f'Discount {discount}% exceeds maximum allowable ceiling ({APPROVAL_DISCOUNT_MAX_PCT}%)',
            fallback_action='CREATE_PAYMENT_LINK'
        )

    # Rule 5: Any discount (above auto-approve ceiling) requires merchant approval
    if discount > AUTO_APPROVE_DISCOUNT_MAX_PCT and discount <= APPROVAL_DISCOUNT_MAX_PCT:
        return PolicyCheck(
            decision=PolicyDecision.NEEDS_MERCHANT_APPROVAL,
            reason=f'Any discount ({discount}%, range {AUTO_APPROVE_DISCOUNT_MAX_PCT}-{APPROVAL_DISCOUNT_MAX_PCT}%) requires merchant approval'
        )

    # Rule 6: Low confidence requires merchant review
    confidence = plan.confidence or 0.0
    if confidence < MIN_CONFIDENCE_AUTO_APPROVE:
        return PolicyCheck(
            decision=PolicyDecision.NEEDS_MERCHANT_APPROVAL,
            reason=f'Planner confidence {confidence:.2f} below auto-approve threshold ({MIN_CONFIDENCE_AUTO_APPROVE})'
        )

    # Rule 7: All checks passed
    return PolicyCheck(
        decision=PolicyDecision.APPROVED,
        reason='Within policy limits'
    )
